### PR TITLE
Adding and integrating new ModelRun artifact for storing training/evaluation logs

### PR DESCRIPTION
**Pull request by 18jeffreyma**
Currently, for Trainer component, GenericExecutor has the user control log writing while Executor writes tensorboard logs in the same directory as model output. We want to have a second artifact to contain model run information (such as Tensorboard logs) seperate from the model output artifact. Changes made:

- Defined ModelRun artifact and added it as an output of Trainer component
- For Trainer GenericExecutor, we add path to ModelRun artifact in the fn_args given to user, so user can write logs to the correct location.
- For Trainer Executor (where logs are written to the same directory as model), we make a copy of the entire serving model directory (which contains logs) to the ModelRun artifact directory
- Keras examples now write Tensorboard output to the ModelRun artifact directory
- Tests added for Trainer component checking that both model and log output is written

### CODE REVIEW

File: tfx/components/trainer/component.py [hunk 1]
```diff
@@ -119,6 +119,7 @@ def __init__(
       custom_config: Optional[Dict[Text, Any]] = None,
       custom_executor_spec: Optional[executor_spec.ExecutorSpec] = None,
       output: Optional[types.Channel] = None,
+      log_output: Optional[types.Channel] = None,
```
Comment (1025KB) [inline tfx/components/trainer/component.py:122]: as it's the run dir / working dir, the usage may not just for tensorboard logs

let's just call it model run or sth generic

File: tfx/components/trainer/component.py [hunk 2]
```diff
@@ -179,6 +180,7 @@ def trainer_fn(trainer.executor.TrainerFnArgs,
         that will be passed into user module.
       custom_executor_spec: Optional custom executor spec.
       output: Optional `Model` channel for result of exported models.
+      log_output: Optional `ModelRun` channel for the log output of models.
```
Comment (1025KB) [inline tfx/components/trainer/component.py:183]: not log only, might sth like 

"Optional `ModelRun` channel, as the working dir of models, can be used to output non-model related output, e.g., tensorboard logs"

File: tfx/components/trainer/executor.py [hunk 3]
```diff
@@ -20,6 +20,7 @@
 
 import json
 import os
+import distutils.dir_util as dir_util
```
Comment (1025KB) [inline tfx/components/trainer/executor.py:23]: use tfx io_utils,

we need to make sure to use tf.io as it's cross platforms

File: tfx/components/trainer/executor.py [hunk 4]
```diff
@@ -278,8 +292,12 @@ def Do(self, input_dict: Dict[Text, List[types.Artifact]],
     tf.estimator.train_and_evaluate(training_spec['estimator'],
                                     training_spec['train_spec'],
                                     training_spec['eval_spec'])
-    absl.logging.info('Training complete.  Model written to %s',
-                      fn_args.serving_model_dir)
+
+    # Copy model run information to ModelRun artifact
+    dir_util.copy_tree(fn_args.serving_model_dir, fn_args.log_dir)
```
Comment (1025KB) [inline tfx/components/trainer/executor.py:297]: only do it if it's chief worker, otherwise it will be messed up in distributed env

File: tfx/components/trainer/executor.py [hunk 5]
```diff
@@ -148,6 +153,8 @@ def _GetFnArgs(self, input_dict: Dict[Text, List[types.Artifact]],
         eval_model_dir=eval_model_dir,
         # A list of uris for eval files.
         eval_files=fn_args.eval_files,
+        # A single uri for the output directory of the Tensorboard logs
+        log_dir=log_output_dir,
```
Comment (1025KB) [inline tfx/components/trainer/executor.py:157]: just call it model_run_dir (or working_dir), sth generic not just logs

File: tfx/examples/airflow_workshop/setup/dags/taxi_utils.py [hunk 6]
```diff
@@ -322,10 +322,9 @@
 #             for i in range(num_dnn_layers)
 #         ])
 #
-#   # TODO(b/158106209): This log path might change in the future.
-#   log_dir = os.path.join(os.path.dirname(fn_args.serving_model_dir), 'logs')
+#   # Write logs to path specfied by fn_args
 #   tensorboard_callback = tf.keras.callbacks.TensorBoard(
-#       log_dir=log_dir, update_freq='batch')
+#       log_dir=fn_args.log_dir, update_freq='batch')
```
Comment (1025KB) [inline tfx/examples/airflow_workshop/setup/dags/taxi_utils.py:327]: just noticed this will make the examples failed with release tfx libraray

we want the examples works with head and release tfx

so let's do 

TODO(b/158106209): deprecate xxx after next release	

if fn_args has log dir: [1]
  log_dir=fn_args.log_dir
else 
  log_dir = os.path.join(os.path.dirname(fn_args.serving_model_dir), 'logs')

[1]not sure if you can just use fn_args.log_dir when fn_args doesn't contains log_dir field in release tfx

File: tfx/examples/mnist/mnist_utils_native_keras_lite.py [hunk 7]
```diff
@@ -77,11 +77,16 @@ def run_fn(fn_args: TrainerFnArgs):
...
+  tensorboard_callback = tf.keras.callbacks.TensorBoard(
+    log_dir=fn_args.log_dir, update_freq='batch')
+
   model.fit(
       train_dataset,
       steps_per_epoch=fn_args.train_steps,
       validation_data=eval_dataset,
-      validation_steps=fn_args.eval_steps)
+      validation_steps=fn_args.eval_steps,
+      callbacks=[tensorboard_callback])
```
Comment (1025KB) [inline tfx/examples/mnist/mnist_utils_native_keras_lite.py:89]: could you try run the e2e test next to those examples to make sure everything works correctly?

File: tfx/utils/path_utils.py [hunk 8]
```diff
@@ -80,3 +81,8 @@ def serving_model_path(output_uri: Text) -> Text:
   else:
     # If dir doesn't match estimator structure, use serving model root directly.
     return serving_model_dir(output_uri)
+
+def log_output_dir(output_uri: Text) -> Text:
```
Comment (1025KB) [inline tfx/utils/path_utils.py:85]: for generic trainer, we just pass the working dir to user, what structure is up to user,

don't need to change path_utils.py

File: tfx/components/trainer/executor.py [hunk 9]
```diff
@@ -278,8 +292,12 @@ def Do(self, input_dict: Dict[Text, List[types.Artifact]],
     tf.estimator.train_and_evaluate(training_spec['estimator'],
                                     training_spec['train_spec'],
                                     training_spec['eval_spec'])
-    absl.logging.info('Training complete.  Model written to %s',
-                      fn_args.serving_model_dir)
+
```
Comment (1025KB) [inline tfx/components/trainer/executor.py:295]: I wonder if there is a way to provide both working_dir and model_dir to trainer_spec, thus we don't need to manually do the split in your next pr (model artifact refactor)

Comment (18jeffreyma) [reply tfx/components/trainer/executor.py:295]: if i recall correcly, the estimator api only writes a model (and its log information) to one directory, but i'll follow up on this in a bit

File: tfx/examples/mnist/mnist_utils_native_keras_lite.py (hunk 7 above)
Comment (18jeffreyma) [reply tfx/examples/mnist/mnist_utils_native_keras_lite.py:94]: e2e tests pass

**Comment by 18jeffreyma** [comment]
Addressed and fixed above issues, except trainer_spec comment

File: tfx/components/trainer/executor.py [hunk 10]
```diff
@@ -206,13 +214,15 @@ def Do(self, input_dict: Dict[Text, List[types.Artifact]],
...
-
+    
     # Note: If trained with multi-node distribution workers, it is the user
     # module's responsibility to export the model only once.
     if not tf.io.gfile.exists(fn_args.serving_model_dir):
       raise RuntimeError('run_fn failed to generate model.')
-    absl.logging.info('Training complete. Model written to %s',
-                      fn_args.serving_model_dir)
+    if not tf.io.gfile.exists(fn_args.model_run_dir):
+      raise RuntimeError('run_fn failed to write logs to correct location.')
```
Comment (1025KB) [inline tfx/components/trainer/executor.py:223]: let's allow empty for now, users may not need tensorboard

in the future we can consider make this an optional output artifact

File: tfx/components/trainer/component.py [hunk 11]
```diff
@@ -226,7 +231,8 @@ def trainer_fn(trainer.executor.TrainerFnArgs,
         run_fn=run_fn,
         trainer_fn=trainer_fn,
         custom_config=json_utils.dumps(custom_config),
-        model=output)
+        model=output,
+        model_run=model_run)
```
Comment (1025KB) [inline tfx/components/trainer/component.py:235]: add a todo here 

`# TODO(b/158106209): change the model_run as optional output artifact`

File: tfx/utils/path_utils.py [hunk 12]
```diff
@@ -79,4 +80,4 @@ def serving_model_path(output_uri: Text) -> Text:
     return io_utils.get_only_uri_in_dir(model_dir)
   else:
     # If dir doesn't match estimator structure, use serving model root directly.
-    return serving_model_dir(output_uri)
+    return serving_model_dir(output_uri)
```
Comment (1025KB) [inline tfx/utils/path_utils.py:83]: keep the last line of the file, otherwise lint will fail, same for other files

File: tfx/utils/path_utils.py [hunk 13]
```diff
@@ -28,6 +28,7 @@
 EVAL_MODEL_DIR = 'eval_model_dir'
 SERVING_MODEL_DIR = 'serving_model_dir'
 
+
```
Comment (1025KB) [inline tfx/utils/path_utils.py:31]: revert the change for this file? or this line is intended?

File: tfx/experimental/templates/taxi/models/keras/model.py [hunk 14]
```diff
@@ -205,11 +205,16 @@ def run_fn(fn_args):
     model = _build_keras_model(
         hidden_units=constants.HIDDEN_UNITS,
         learning_rate=constants.LEARNING_RATE)
-  # This log path might change in the future.
+
+   # TODO(b/158106209): deprecate logging to model artifact after next release
   log_dir = os.path.join(os.path.dirname(fn_args.serving_model_dir), 'logs')
+  if (fn_args.model_run_dir):
```
Comment (1025KB) [inline tfx/experimental/templates/taxi/models/keras/model.py:211]: did you check in tfx the release/old version (doesn't have model_run_dir defined when constructing TrainerFnArgs, not even set to None), will fn_args.model_run_dir be none or error?

File: tfx/examples/mnist/mnist_utils_native_keras_lite.py (hunk 7 above)
Comment (1025KB) [reply tfx/examples/mnist/mnist_utils_native_keras_lite.py:94]: could you also run a e2e example, and check the ~/tfx outputs and see if tensorboard works correctly

File: tfx/experimental/templates/taxi/models/keras/model.py (hunk 14 above)
Comment (18jeffreyma) [reply tfx/experimental/templates/taxi/models/keras/model.py:211]: i think i remember seeing that this worked, but wasn't sure where. I've rewritten it to catch a key error instead

**Comment by 18jeffreyma** [comment]
Addressed all new changes in refactoring and in the KeyErrors what would come up. Tensorboard logs are correctly written for the mnist e2e example in keras.

File: tfx/components/trainer/executor.py [hunk 15]
```diff
@@ -178,6 +185,7 @@ def Do(self, input_dict: Dict[Text, List[types.Artifact]],
         - schema: Schema of the data.
       output_dict: Output dict from output key to a list of Artifacts.
         - output: Exported model.
+        - model_run: Model trainng related outputs (e.g., Tensorboard logs)
```
Comment (1025KB) [inline tfx/components/trainer/executor.py:188]: could you also update above -output to -model (we changed key from `output` to `model` but forget to update here )

File: tfx/components/trainer/executor.py [hunk 16]
```diff
@@ -206,13 +214,14 @@ def Do(self, input_dict: Dict[Text, List[types.Artifact]],
...
-
+    
     # Note: If trained with multi-node distribution workers, it is the user
     # module's responsibility to export the model only once.
     if not tf.io.gfile.exists(fn_args.serving_model_dir):
       raise RuntimeError('run_fn failed to generate model.')
-    absl.logging.info('Training complete. Model written to %s',
-                      fn_args.serving_model_dir)
+    
+    absl.logging.info('Training complete. Model written to %s. Logs written to %s',
```
Comment (1025KB) [inline tfx/components/trainer/executor.py:223]: Logs -> ModelRun

File: tfx/components/trainer/executor.py [hunk 17]
```diff
@@ -245,6 +256,7 @@ def Do(self, input_dict: Dict[Text, List[types.Artifact]],
         - schema: Schema of the data.
       output_dict: Output dict from output key to a list of Artifacts.
         - output: Exported model.
+        - model_run: Model trainng related outputs (e.g., Tensorboard logs)
```
Comment (1025KB) [inline tfx/components/trainer/executor.py:259]: -output to - model

File: tfx/components/trainer/executor.py [hunk 18]
```diff
@@ -224,7 +233,9 @@ class Executor(GenericExecutor):
   https://github.com/tensorflow/tfx/blob/master/tfx/examples/chicago_taxi_pipeline/taxi_utils.py#L285.
   This becomes the basis of the new Executor for Trainer. This Executor will
   then train and evaluate this estimator using the
-  tf.estimator.train_and_evaluate API to train locally.
+  tf.estimator.train_and_evaluate API to train locally, will save the
+  moel to the desired ouput artifact, and will copy the training logs to a 
+  seperate ModelRun artifact.
```
Comment (1025KB) [inline tfx/components/trainer/executor.py:238]: I feel we can just keep the previous high level comment here

File: tfx/components/trainer/executor.py [hunk 19]
```diff
@@ -290,10 +303,14 @@ def Do(self, input_dict: Dict[Text, List[types.Artifact]],
...
+      io_utils.copy_dir(fn_args.serving_model_dir, fn_args.model_run_dir)
+
       absl.logging.info('Exported eval_savedmodel to %s.',
                         fn_args.eval_model_dir)
     else:
       absl.logging.info(
           'eval_savedmodel export for TFMA is skipped because '
           'this is not the chief worker.'
       )
+
```
Comment (1025KB) [inline tfx/components/trainer/executor.py:316]: revert this diff?

File: tfx/experimental/templates/taxi/models/keras/model.py (hunk 14 above)
Comment (1025KB) [reply tfx/experimental/templates/taxi/models/keras/model.py]: you can test it print(fn_args.not_exist_name == None)

btw, I didn't see the change, did you commit?

File: tfx/components/trainer/executor.py (hunk 19 above)
Comment (18jeffreyma) [reply tfx/components/trainer/executor.py:316]: i added this for the linter (which you mentioned in a previous comment)

**Comment by 18jeffreyma** [comment]
Addressed all new comments; for the last comment, I think you were looking at an outdated commit, the changed file is pushed.

File: tfx/components/trainer/executor.py (hunk 19 above)
Comment (1025KB) [reply tfx/components/trainer/executor.py:314]: I mean the file ends with a \n

basically here shouldn't be any diff compare to original executor.py

File: tfx/examples/airflow_workshop/setup/dags/taxi_utils_solution.py [hunk 20]
```diff
@@ -319,8 +319,14 @@ def run_fn(fn_args: TrainerFnArgs):
             for i in range(num_dnn_layers)
         ])
 
-  # TODO(b/158106209): This log path might change in the future.
+  # TODO(b/158106209): deprecate logging to model artifact after next release
   log_dir = os.path.join(os.path.dirname(fn_args.serving_model_dir), 'logs')
+  try:
+    log_dir = fn_args.model_run_dir
+  except KeyError:
+    pass
```
Comment (1025KB) [inline tfx/examples/airflow_workshop/setup/dags/taxi_utils_solution.py:327]:
```
try:
  log_dir = fn_args.model_run_dir
except:
  # TODO(b/xxx): deprecate xxx after next release.
  log_dir = os.path.join(os.path.dirname(fn_args.serving_model_dir), 'logs')
```

**Comment by 18jeffreyma** [comment]
addressed most recent round of comments

File: tfx/components/trainer/executor.py [hunk 21]
```diff
@@ -177,7 +184,8 @@ def Do(self, input_dict: Dict[Text, List[types.Artifact]],
         - transform_output: Optional input transform graph.
         - schema: Schema of the data.
       output_dict: Output dict from output key to a list of Artifacts.
-        - output: Exported model.
+        - model: Exported model.
+        - model_run: Model trainng related outputs (e.g., Tensorboard logs)
```
Comment (goutham) [inline tfx/components/trainer/executor.py:188]: typo: trainng -> training

File: tfx/experimental/templates/taxi/models/keras/model.py [hunk 22]
```diff
@@ -205,11 +205,17 @@ def run_fn(fn_args):
...
+    log_dir = fn_args.model_run_dir
+  except KeyError:
+    # TODO(b/158106209): deprecate logging to model artifact after next release
+    log_dir = os.path.join(os.path.dirname(fn_args.serving_model_dir), 'logs')
+
+  # Write logs to path
   tensorboard_callback = tf.keras.callbacks.TensorBoard(
       log_dir=log_dir, update_freq='batch')
-
+
```
Comment (goutham) [inline tfx/experimental/templates/taxi/models/keras/model.py:218]: Remove unnecessary whitespace.

File: tfx/types/standard_artifacts.py [hunk 23]
```diff
@@ -81,6 +81,10 @@ class Model(Artifact):
   TYPE_NAME = 'Model'
 
 
+class ModelRun(Artifact):
+  TYPE_NAME = "ModelRun"
```
Comment (goutham) [inline tfx/types/standard_artifacts.py:85]: Use single quotes to be consistent.

**Comment by 18jeffreyma** [comment]
addressed all recently added issues, thanks!
